### DISCUSSION TITLE
What's lsp-deferred?

### DISCUSSION THREAD

**Discussion by gzagatti** [emacs-lsp/lsp-mode#3360] [Q&A] [answered]
LSP is working fine with my configuration, but as I'm trying to tweak the configuration to suit my needs I came across the command `lsp-deferred` in [the installation page](https://emacs-lsp.github.io/lsp-mode/page/installation/). It says:

> To defer LSP server startup (and DidOpen notifications) until the buffer is visible you can use lsp-deferred instead of lsp:
>
> ```lisp
> (add-hook 'XXX-mode-hook #'lsp-deferred)
> ```

What do you mean by defer the server startup until the buffer is visible? What happens when it is not deferred?

**Comment by kiennq** [comment, marked as answer]
> What do you mean by defer the server startup until the buffer is visible? What happens when it is not deferred?

If it's not deferred, the server starts immediately when the `lsp` command is encountered.
That can take a very long time when you open several buffers at one and have `lsp-mode` starts in all of them. You can even feel like Emacs is hanging.

An example is when you restore from `desktop-save-mode` and have a dozen buffers restored, you may not want to have `lsp-mode` running in all of the buffers immediately, but only when they're visible in your view.

**Reply by gzagatti** [reply]
Thanks so much for the detailed response.

**Reply by trev-dev** [reply]
I honestly don't understand the distinction. Why would you prefer to use 'lsp if 'lsp-deferred is just...safer?

**Reply by kiennq** [reply]
LSP starting/indexing takes time. Sometimes, when you can afford long Emacs startups, you might want to just start lsp servers even when the buffers are not visible, so when you need it, it immediately available to use.
